nat_unix.go

```diff
@@ -0,0 +1,28 @@
+// +build !windows
+
+package nat
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+)
+
+func setTOS(conn net.Conn, tos int) error {
+	if tos <= 0 {
+		return nil
+	}
+	ipConn, ok := conn.(*net.IPConn)
+	if !ok {
+		return errors.New("Failed to set TOS, connection is not an IPConn")
+	}
+	f, err := ipConn.File()
+	if err != nil {
+		return fmt.Errorf("Failed to set TOS, connection retrieve the fd: %v", err)
+	}
+	if err := syscall.SetsockoptInt(int(f.Fd()), syscall.SOL_SOCKET, syscall.IP_TOS, tos); err != nil {
```

Comment:
I think you can use IPConn.SyscallConn() and then use RawConn.Control (https://golang.org/pkg/syscall/#RawConn) to get portable access to the underlying file descriptor. Doesn't that let you avoid per-OS settings?

Reply:
I'm not sure how. From a code POV, RawConn Control still uses an uniptr which has to be casted to different types depending on the architecture. More fundamentally, pkg/syscall is different code depending on the GOARCH. I think this is the main reason why it is deprecatish for golang.org/x/sys/ which has the variant unix, windows and plan9 which can live together.
If you want I can stuff everything in a single file by using the mentioned golang.org/x/sys/unix and golang.org/x/sys/windows and have a run-time conditional based on runtime.GOARCH. I personally prefer my first solution because it link a bit less of code, but I am OK with the second as well. Just let me know.